apps/meteor/ee/server/hooks/messages/BeforeSaveCannedResponse.ts

```diff
@@ -0,0 +1,116 @@
+import { isILivechatVisitor, isOmnichannelRoom } from '@rocket.chat/core-typings';
+import type { IMessage, IRoom, ISetting, IUser, SettingValue } from '@rocket.chat/core-typings';
+import { License } from '@rocket.chat/license';
+import { LivechatVisitors, Users } from '@rocket.chat/models';
+import get from 'lodash.get';
+import mem from 'mem';
+
+const placeholderFields = {
+	'contact.name': {
+		from: 'visitor',
+		dataKey: 'name',
+	},
+	'contact.email': {
+		from: 'visitor',
+		dataKey: 'visitorEmails[0].address',
+	},
+	'contact.phone': {
+		from: 'visitor',
+		dataKey: 'phone[0].phoneNumber',
+	},
+	'agent.name': {
+		from: 'agent',
+		dataKey: 'name',
+	},
+	'agent.email': {
+		from: 'agent',
+		dataKey: 'emails[0].address',
+	},
+};
+
+export class BeforeSaveCannedResponse {
+	static enabled = false;
+
+	private getSetting: <T extends SettingValue = SettingValue>(id: ISetting['_id']) => T;
+
+	constructor({ getSetting }: { getSetting: <T extends SettingValue = SettingValue>(id: ISetting['_id']) => T }) {
```

Comment:
type :trollface: 

Was thinking what's better, having to type the object or initializing it as predicate read-only in the constructor 🤔

Reply:
I don't know as well.. I wanted to decouple this as much as possible from the rest, but I couldn't decouple from the setting so I decided to go this way but definitely not something I'm happy with.

maybe here I can have only the static property to say it is enabled or not and control it from outside, like having the setting and license listener somewhere else in the omnichannel service.. wdyt?

Reply:
That is def an option, and will help with the testability 🤔

What if all "callbacks" have this static prop of "enabled", like you suggest, with a setter that prints a debug log when the value is changed? That way we can at least track why a new callback is not working, while decoupling that event logic from here 👀

Reply:
oh, I ended up not creating the setter nor adding a log.. but changed to rely only on the `enabled` flag